### PR TITLE
core: Add field to track index of first message in each file split.

### PR DESCRIPTION
**Pull request by haiqi96**
# References

# Description

This PR is motivated by the need of filtering a file split with a specific message index.

CLP splits input file when the file size is greater than a threshold. Before this change, there is no simple way to know the range of message index in each file split. The information can be inferred by ordering the file split in the ascending creation order and summing up the num_messages in each file split, which is inefficient and complicated.

This change adds a begin_msg_idx to both Archive and Global metadata database, so the value can be looked up easily from the database.

# Validation performed
Compressed and decompressed test input with CLP and CLP-S, ensured there is no error.
Manually examined the begin message offset in the database and confirmed their values are expected.

Tested with running a global MySQL database with the package and verified that the CLP can compress with using a db-config file

### CODE REVIEW

File: components/core/src/clp/streaming_archive/writer/File.hpp [hunk 1]
```diff
@@ -104,6 +105,24 @@ class File {
...
     bool has_ts_pattern() const { return m_timestamp_patterns.empty() == false; }
 
+    /**
+     * Sets combined file message offset
+     * @param offset
+     */
+    void set_combined_file_msg_offset(size_t offset);
+
+    /**
+     * Gets combined file message offset
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/clp/streaming_archive/writer/File.hpp:115]: I think we can omit the `Gets` description for pure getters and just use `@return`, even when it's inconsistent with the rest of the file. The idea being that we don't want to create extra work for ourselves in refactoring the file (and comments are low-impact places to be inconsistent). We could submit a PR right after this to make the docstrings consistent in this file.

File: components/clp-py-utils/clp_py_utils/initialize-clp-metadata-db.py [hunk 2]
```diff
@@ -84,6 +84,7 @@ def main(argv):
                     `path` VARCHAR(12288) NOT NULL,
                     `begin_timestamp` BIGINT NOT NULL,
                     `end_timestamp` BIGINT NOT NULL,
+                    `begin_msg_idx` BIGINT NOT NULL,
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/clp-py-utils/clp_py_utils/initialize-clp-metadata-db.py:87]: This doesn't affect functionality, but it's nicer to be consistent about the ordering of columns here and in the other code---i.e., put `begin_message_ix` before `num_messages`.

File: components/core/src/clp/streaming_archive/Constants.hpp [hunk 3]
```diff
@@ -38,6 +38,7 @@ constexpr char BeginTimestamp[] = "begin_timestamp";
 constexpr char EndTimestamp[] = "end_timestamp";
 constexpr char TimestampPatterns[] = "timestamp_patterns";
 constexpr char NumUncompressedBytes[] = "num_uncompressed_bytes";
+constexpr char BeginMsgIdx[] = "begin_msg_idx";
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/Constants.hpp:41]: Ditto

File: components/core/tools/scripts/db/init-db.py [hunk 4]
```diff
@@ -84,6 +84,7 @@ def main(argv):
                 `path` VARCHAR(12288) NOT NULL,
                 `begin_timestamp` BIGINT NOT NULL,
                 `end_timestamp` BIGINT NOT NULL,
+                `begin_msg_idx` BIGINT NOT NULL,
```
Comment (kirkrodrigues) [inline components/core/tools/scripts/db/init-db.py:87]: Ditto

File: components/core/src/clp/streaming_archive/reader/File.cpp [hunk 5]
```diff
@@ -32,6 +32,7 @@ ErrorCode File::open_me(
     file_metadata_ix.get_path(m_orig_path);
     m_begin_ts = file_metadata_ix.get_begin_ts();
     m_end_ts = file_metadata_ix.get_end_ts();
+    m_begin_msg_idx = file_metadata_ix.get_begin_msg_idx();
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/reader/File.cpp:35]: I would move this above just above where `m_num_messages` is set.

File: components/core/src/clp/streaming_archive/writer/File.hpp [hunk 6]
```diff
@@ -104,6 +105,22 @@ class File {
      */
     bool has_ts_pattern() const { return m_timestamp_patterns.empty() == false; }
 
+    /**
+     * Sets the File's begin message index
+     * @param offset
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/writer/File.hpp:110]:
```suggestion
     * @param msg_idx
```

File: components/core/src/clp/streaming_archive/writer/Archive.hpp [hunk 7]
```diff
@@ -128,6 +128,13 @@ class Archive {
      */
     void set_file_is_split(bool is_split);
 
+    /**
+     * Wrapper for streaming_archive::writer::File::set_begin_msg_idx
+     * @param file File to set the combined-file message msg_idx for
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/writer/Archive.hpp:133]:
```suggestion
     * @param file
```

File: components/core/src/clp/streaming_archive/writer/utils.cpp [hunk 8]
```diff
@@ -49,16 +51,18 @@ auto split_file_and_archive(
     auto has_ts_pattern = encoded_file.has_ts_pattern();
     auto orig_file_id = encoded_file.get_orig_file_id();
     auto split_ix = encoded_file.get_split_ix();
+    auto const end_msg_idx = encoded_file.get_end_msg_idx();
```
Comment (kirkrodrigues) [inline components/core/src/clp/streaming_archive/writer/utils.cpp:54]: If we're making the variables in `split_file` `const`, let's do the same here for consistency.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/src/clp/streaming_archive/writer/utils.cpp [hunk 9]
```diff
@@ -25,17 +25,19 @@ auto split_file(
...
     archive_writer.set_file_is_split(true);
     close_file_and_append_to_segment(archive_writer);
 
-    archive_writer.create_and_open_file(path_for_compression, group_id, orig_file_id, ++split_ix);
+    archive_writer.create_and_open_file(path_for_compression, group_id, orig_file_id, split_ix + 1);
     if (has_ts_pattern) {
         // Initialize the file's timestamp pattern to the previous split's pattern
         archive_writer.change_ts_pattern(last_timestamp_pattern);
     }
+    archive_writer.set_file_begin_message_ix(end_message_ix);
```
Comment (LinZhihao-723) [inline components/core/src/clp/streaming_archive/writer/utils.cpp:40]: I think the begin message index should be passed in whenever a file is created and we should set it after creation. Let's make it as a parameter in `create_and_open_file`

**Review by kirkrodrigues — APPROVED** [review]
How about:

core: Add field to track index of first message in each file split. (#412)

**Review by LinZhihao-723 — APPROVED** [review]

**Pull request merged**